File: vulnerable_app.ts
import * as fs from 'fs';
import * as crypto from 'crypto';
import { exec } from 'child_process';
import * as http from 'http';

// Hardcoded credentials (CWE-798)
const DB_PASSWORD: string = "ts_hardcoded_pass_123";
const JWT_SECRET: string = "ts_jwt_secret_key_never_rotate";
const PRIVATE_KEY: string = "-----BEGIN RSA PRIVATE KEY-----\nhardcodedkey\n-----END RSA PRIVATE KEY-----";

// SQL Injection (CWE-89)
function getUser(db: any, username: string): any {
    return db.query(`SELECT * FROM users WHERE username = '${username}'`);
}

function searchProducts(db: any, term: string): any {
    return db.query("SELECT * FROM products WHERE name LIKE '%" + term + "%'");
}

// Command Injection (CWE-78)
function pingHost(host: string): void {
    exec(`ping -c 1 ${host}`, (err, stdout) => console.log(stdout));
}

function runScript(script: string): void {
    exec(script, (err, stdout) => console.log(stdout));
}

// Path Traversal (CWE-22)
function readFile(filename: string): string {
    return fs.readFileSync(`/var/app/${filename}`, 'utf8');
}

function serveFile(req: any, res: any): void {
    const path = req.query.path;
    res.send(fs.readFileSync(`/uploads/${path}`));
}

// Weak cryptography (CWE-326)
function hashPassword(password: string): string {
    return crypto.createHash('md5').update(password).digest('hex');
}

function signData(data: string): string {
    return crypto.createHash('sha1').update(data).digest('hex');
}

// Insecure random (CWE-330)
function generateToken(): string {
    return Math.random().toString(36).substring(2);
}

function generateOTP(): number {
    return Math.floor(Math.random() * 9000) + 1000;
}

// XSS (CWE-79)
function renderPage(userInput: string): string {
    return `<html><body>${userInput}</body></html>`;
}

function greetUser(name: string): string {
    return `<h1>Hello ${name}</h1>`;
}

// SSRF (CWE-918)
function fetchURL(req: any, res: any): void {
    const url = req.query.url;
    http.get(url, (response: any) => {
        let data = '';
        response.on('data', (chunk: string) => data += chunk);
        response.on('end', () => res.send(data));
    });
}

// eval injection (CWE-95)
function calculate(expression: string): any {
    return eval(expression);
}

// Prototype pollution
function mergeObjects(target: any, source: any): any {
    for (const key in source) {
        target[key] = source[key];
    }
    return target;
}

// Open redirect (CWE-601)
function redirectUser(res: any, url: string): void {
    res.redirect(url);
}

// Insecure JWT (none algorithm) (CWE-347)
function createToken(payload: object): string {
    const header = Buffer.from(JSON.stringify({ alg: 'none', typ: 'JWT' })).toString('base64');
    const body = Buffer.from(JSON.stringify(payload)).toString('base64');
    return `${header}.${body}.`;
}

// Sensitive data in logs (CWE-532)
function login(user: string, password: string): boolean {
    console.log(`Login: user=${user} password=${password}`);
    return password === DB_PASSWORD;
}

// ReDoS (CWE-1333)
function validateEmail(email: string): boolean {
    return /^([a-zA-Z0-9]+)*@[a-zA-Z0-9]+\.[a-zA-Z]+$/.test(email);
}

File: vulnerable_app.go
package main

import (
	"crypto/md5"
	"database/sql"
	"fmt"
	"io/ioutil"
	"math/rand"
	"net/http"
	"os"
	"os/exec"
)

// Hardcoded credentials (CWE-798)
const DBPassword = "go_hardcoded_pass_123"
const AWSKey = "AKIAIOSFODNN7GOEXAMPLE"
const AWSSecret = "wJalrXUtnFEMI/K7MDENG/bPxRfiCYGOEXAMPLE"

// SQL Injection (CWE-89)
func getUser(db *sql.DB, username string) {
	query := "SELECT * FROM users WHERE username = '" + username + "'"
	db.Query(query)
}

func getOrder(db *sql.DB, orderID string) {
	db.Query(fmt.Sprintf("SELECT * FROM orders WHERE id = %s", orderID))
}

// Command Injection (CWE-78)
func pingHost(host string) {
	exec.Command("sh", "-c", "ping -c 1 "+host).Run()
}

func runScript(script string) {
	cmd := exec.Command("/bin/bash", "-c", script)
	cmd.Run()
}

// Path Traversal (CWE-22)
func readFile(filename string) ([]byte, error) {
	return ioutil.ReadFile("/var/app/" + filename)
}

func serveFile(w http.ResponseWriter, r *http.Request) {
	path := r.URL.Query().Get("file")
	data, _ := os.ReadFile("/uploads/" + path)
	w.Write(data)
}

// Weak cryptography (CWE-326)
func hashPassword(password string) string {
	h := md5.New()
	h.Write([]byte(password))
	return fmt.Sprintf("%x", h.Sum(nil))
}

// Insecure random (CWE-330)
func generateToken() int {
	return rand.Intn(999999)
}

func generateOTP() string {
	return fmt.Sprintf("%04d", rand.Intn(10000))
}

// SSRF (CWE-918)
func fetchURL(w http.ResponseWriter, r *http.Request) {
	url := r.URL.Query().Get("url")
	resp, _ := http.Get(url)
	body, _ := ioutil.ReadAll(resp.Body)
	w.Write(body)
}

// XSS (CWE-79)
func greetUser(w http.ResponseWriter, r *http.Request) {
	name := r.URL.Query().Get("name")
	fmt.Fprintf(w, "<h1>Hello %s</h1>", name)
}

// Sensitive data in logs (CWE-532)
func login(user, password string) bool {
	fmt.Printf("Login attempt: user=%s password=%s\n", user, password)
	return password == DBPassword
}

// Open redirect (CWE-601)
func redirectHandler(w http.ResponseWriter, r *http.Request) {
	url := r.URL.Query().Get("url")
	http.Redirect(w, r, url, http.StatusFound)
}

// Nil pointer dereference (CWE-476)
func getLength(s *string) int {
	return len(*s)
}

func main() {
	http.HandleFunc("/fetch", fetchURL)
	http.HandleFunc("/greet", greetUser)
	http.HandleFunc("/redirect", redirectHandler)
	http.HandleFunc("/file", serveFile)
	http.ListenAndServe(":8080", nil)
}
